#### ETL SOBRE STEAM_GAMES

In [3]:
import pandas as pd
import json
import ast
import warnings
import pyarrow as pa
import pyarrow.parquet as pq
warnings.filterwarnings('ignore')

In [4]:
# Se carga el json de steam_games y se realiza una lectura linea a linea que se irá almacenando en una lista para luego convertirla en un dataframe pandas
steam_games = 'datos_json/output_steam_games.json'

listado = []

with open(steam_games) as archivo:
    for linea in archivo.readlines():
        fila = json.loads(linea)
        listado.append(fila)

dfSteamGames = pd.DataFrame(listado)

# Se revisa la cantidad de nulos por columna en el archivo
dfSteamGames.isnull().sum()

publisher       96362
genres          91593
app_name        88312
title           90360
url             88310
release_date    90377
tags            88473
reviews_url     88312
specs           88980
price           89687
early_access    88310
id              88312
developer       91609
dtype: int64

In [5]:
# Se limpian las filas donde todos los valores sean nulos y se reinicia el index 
games = dfSteamGames.dropna(how='all').reset_index(drop=True)

games

publisher                                             genres  \
0             Kotoshiro      [Action, Casual, Indie, Simulation, Strategy]   
1      Making Fun, Inc.               [Free to Play, Indie, RPG, Strategy]   
2          Poolians.com  [Casual, Free to Play, Indie, Simulation, Sports]   
3                  彼岸领域                        [Action, Adventure, Casual]   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games              [Casual, Indie, Simulation, Strategy]   
32131            Sacada                          [Casual, Indie, Strategy]   
32132      Laush Studio                        [Indie, Racing, Simulation]   
32133          SIXNAILS                                    [Casual, Indie]   
32134               NaN                                                NaN   

                       app_name                     title  \
0           Lost Summoner Kitty       Lost Summoner Kitty   
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  \
0          [Strategy, Action, Indie, Casual, Simulation]   
1      [Free to Play, Strategy, Indie, RPG, Card Game...   
2      [Free to Play, Simulation, Sports, Casual, Ind...   
3                            [Action, Adventure, Casual]   
4                        [Action, Indie, Casual, Sports]   
...                                                  ...   
32130              [Strategy, Indie, Casual, Simulation]   
32131                          [Strategy, Indie, Casual]   
32132                        [Indie, Simulation, Racing]   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
32134  [Early Access, Adventure, Indie, Action, Simul...   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                         

In [6]:
# Se aplica el metodo drop_duplicates utilizando como filtro el id
games = games.drop_duplicates(subset='id')

In [7]:
'''  
OBSERVACIONES

id y price estan como objeto, debería transformarlos a int
release_date esta como objeto, podria transformarse a dato tipo date?
genres, tags y specs tienen datos anidados, deberia desanidarlos?


'''
games.dtypes

publisher       object
genres          object
app_name        object
title           object
url             object
release_date    object
tags            object
reviews_url     object
specs           object
price           object
early_access    object
id              object
developer       object
dtype: object

In [8]:
# Se desanida la columna genres para poder tener cada genero en una fila
steam_games = games.explode('genres')

# Se reindexa el dataframe luego de desanidarse para su correcto acceso a la información
steam_games = steam_games.reset_index(drop=True)

steam_games.head()

publisher      genres             app_name                title  \
0  Kotoshiro      Action  Lost Summoner Kitty  Lost Summoner Kitty   
1  Kotoshiro      Casual  Lost Summoner Kitty  Lost Summoner Kitty   
2  Kotoshiro       Indie  Lost Summoner Kitty  Lost Summoner Kitty   
3  Kotoshiro  Simulation  Lost Summoner Kitty  Lost Summoner Kitty   
4  Kotoshiro    Strategy  Lost Summoner Kitty  Lost Summoner Kitty   

                                                 url release_date  \
0  http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1  http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
2  http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
3  http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
4  http://store.steampowered.com/app/761140/Lost_...   2018-01-04   

                                            tags  \
0  [Strategy, Action, Indie, Casual, Simulation]   
1  [Strategy, Action, Indie, Casual, Simulation]   
2  [Strategy, Action, Indie, Casual, Simulation]   
3  [Strategy, Action, Indie, Casual, Simulation]   
4  [Strategy, Action, Indie, Casual, Simulation]   

                                         reviews_url            specs price  \
0  http://steamcommunity.com/app/761140/reviews/?...  [Single-player]  4.99   
1  http://steamcommunity.com/app/761140/reviews/?...  [Single-player]  4.99   
2  http://steamcommunity.com/app/761140/reviews/?...  [Single-player]  4.99   
3  http://steamcommunity.com/app/761140/reviews/?...  [Single-player]  4.99   
4  http://steamcommunity.com/app/761140/reviews/?...  [Single-player]  4.99   

  early_access      id  developer  
0        False  761140  Kotoshiro  
1        False  761140  Kotoshiro  
2        False  761140  Kotoshiro  
3        False  761140  Kotoshiro  
4        False  761140  Kotoshiro

In [9]:
steam_games.isnull().sum()

publisher       15594
genres           3283
app_name            3
title            2051
url                 0
release_date     2102
tags              185
reviews_url         1
specs             941
price            3346
early_access        0
id                  1
developer        3479
dtype: int64

In [10]:
# Almaceno los indices de los nulos en la columna genres que no me permiten realizar la transformacion del tipo de dato
generos_nulos = steam_games[steam_games['genres'].isnull()].index


# Utilizo la variable nulos donde almacene el indice de nulos para ubicar las filas correspondientes
steam_games.iloc[generos_nulos]


publisher genres                                    app_name  \
17          NaN    NaN                               Log Challenge   
41          NaN    NaN                        Icarus Six Sixty Six   
70          NaN    NaN                               After Life VR   
71          NaN    NaN                                  Kitty Hawk   
75          NaN    NaN                                  Mortars VR   
...         ...    ...                                         ...   
74796       NaN    NaN   Robotpencil Presents: Exercise: Brushwork   
74797       NaN    NaN  Robotpencil Presents: Creative Composition   
74798       NaN    NaN                            The Gamble House   
74799       NaN    NaN    Kalen Chock Presents: 2017 Free Tutorial   
74833       NaN    NaN                                 Maze Run VR   

                                            title  \
17                                            NaN   
41                                            NaN   
70                                            NaN   
71                                            NaN   
75                                            NaN   
...                                           ...   
74796   Robotpencil Presents: Exercise: Brushwork   
74797  Robotpencil Presents: Creative Composition   
74798                            The Gamble House   
74799    Kalen Chock Presents: 2017 Free Tutorial   
74833                                         NaN   

                                                     url release_date  \
17     http://store.steampowered.com/app/773570/Log_C...          NaN   
41     http://store.steampowered.com/app/724910/Icaru...          NaN   
70     http://store.steampowered.com/app/772590/After...          NaN   
71     http://store.steampowered.com/app/640250/Kitty...          NaN   
75     http://store.steampowered.com/app/711440/Morta...          NaN   
...                                                  ...          ...   
74796  http://store.steampowered.com/app/775640/Robot...   2018-01-03   
74797  http://store.steampowered.com/app/777930/Robot...   2018-01-03   
74798  http://store.steampowered.com/app/775370/The_G...   2016-11-19   
74799  http://store.steampowered.com/app/777950/Kalen...   2018-01-03   
74833  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  \
17                       [Action, Indie, Casual, Sports]   
41                                              [Casual]   
70                             [Early Access, Indie, VR]   
71      [Early Access, Action, Adventure, Indie, Casual]   
75     [Early Access, Strategy, Action, Indie, Casual...   
...                                                  ...   
74796                  [Design & Illustration, Tutorial]   
74797                  [Design & Illustration, Tutorial]   
74798                                            [Movie]   
74799                  [Design & Illustration, Tutorial]   
74833  [Early Access, Adventure, Indie, Action, Simul...   

                                             reviews_url  \
17     http://steamcommunity.com/app/773570/reviews/?...   
41     http://steamcommunity.com/app/724910/reviews/?...   
70     http://steamcommunity.com/app/772590/reviews/?...   
71     http://steamcommunity.com/app/640250/reviews/?...   
75     http://steamcommunity.com/app/711440/reviews/?...   
...                                                  ...   
74796  http://steamcommunity.com/app/775640/reviews/?...   
74797  http://steamcommunity.com/app/777930/reviews/?...   
74798  http://steamcommunity.com/app/775370/reviews/?...   
74799  http://steamcommunity.com/app/777950/reviews/?...   
74833  http://steamcommunity.com/app/681550/reviews/?...   

                                                   specs price early_access  \
17     [Single-player, Full controller support, HTC V...  2.99        False   
41     [Single-player, HTC Vive, Tracked Motion Con

#### En base a lo observado:

- hay 3283 registros con gran cantidad de nulos, al analizar en profundidad hay columnas que repiten información, se elige aplicar el metodo fillna para rellenar las columnas nulas con las columnas que tienen los datos necesarios. 

In [11]:
# Se interpola la columna title con la columna app_name para rellenar los nulos utilizando el metodo fillna
steam_games['title'] = steam_games['title'].fillna(steam_games['app_name'])

In [12]:
# Almaceno los indices de los nulos en la columna id que no me permiten realizar la transformacion del tipo de dato
id_nulos = steam_games[steam_games['id'].isna()].index

# Utilizo la variable nulos donde almacene el indice de nulos para ubicar las filas correspondientes
steam_games.iloc[id_nulos]

publisher genres app_name title                             url  \
189       NaN    NaN      NaN   NaN  http://store.steampowered.com/   

    release_date tags reviews_url specs  price early_access   id developer  
189          NaN  NaN         NaN   NaN  19.99        False  NaN       NaN

#### En base a lo observado:

- la fila 189 va a ser eliminada ya que un 80% de sus datos son nulos

In [13]:
# Se elimina la fila 189
steam_games.drop(index=189,inplace=True)

#### 

In [14]:
# Se aplica la transformación del tipo en la columna id
steam_games['id'] = steam_games['id'].astype(int)

In [15]:
# Se cambia la columna a tipo numero y los valores que sean str los convierte en nulos
steam_games["price"] = pd.to_numeric(steam_games["price"], errors='coerce')

# Se rellenan los nulos de la columna con 0
steam_games["price"] = steam_games["price"].fillna(0)

In [16]:
# Se cambia el tipo de la columna release_date a datetime
steam_games["release_date"] = pd.to_datetime(steam_games["release_date"],errors='coerce')

# Se extrae el año de la columna release_date y se lo almacena en una nueva columna
steam_games["year"] = steam_games["release_date"].dt.year

In [17]:
# Se aplica una interpolación cruzada a los datos nulos que hayan quedado en la columna year
steam_games["year"].interpolate(method='ffill', inplace=True)

In [18]:
# Se cambia el tipo de la columna year a int
steam_games['year'] = steam_games['year'].astype(int)

In [19]:
steam_games.dropna(thresh= 4, inplace= True)

In [20]:
steam_games.dtypes

publisher               object
genres                  object
app_name                object
title                   object
url                     object
release_date    datetime64[ns]
tags                    object
reviews_url             object
specs                   object
price                  float64
early_access            object
id                       int32
developer               object
year                     int32
dtype: object

In [21]:
steam_games.isna().mean()

publisher       0.208371
genres          0.043858
app_name        0.000027
title           0.000027
url             0.000000
release_date    0.038726
tags            0.002459
reviews_url     0.000000
specs           0.012561
price           0.000000
early_access    0.000000
id              0.000000
developer       0.046477
year            0.000000
dtype: float64

In [22]:
# Se almacenan las columnas para aplicar el metodo lower
columnas_a_minusculas = ['publisher', 'genres', 'app_name','title','developer']

# Se lleva a minusculas las columnas previamente almacenadas
steam_games[columnas_a_minusculas] = steam_games[columnas_a_minusculas].apply(lambda x: x.str.lower())

In [23]:
steam_games

publisher      genres             app_name                title  \
0         kotoshiro      action  lost summoner kitty  lost summoner kitty   
1         kotoshiro      casual  lost summoner kitty  lost summoner kitty   
2         kotoshiro       indie  lost summoner kitty  lost summoner kitty   
3         kotoshiro  simulation  lost summoner kitty  lost summoner kitty   
4         kotoshiro    strategy  lost summoner kitty  lost summoner kitty   
...             ...         ...                  ...                  ...   
74829  laush studio      racing        russian roads        russian roads   
74830  laush studio  simulation        russian roads        russian roads   
74831      sixnails      casual  exit 2 - directions  exit 2 - directions   
74832      sixnails       indie  exit 2 - directions  exit 2 - directions   
74833           NaN         NaN          maze run vr          maze run vr   

                                                     url release_date  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
2      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
3      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
4      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
...                                                  ...          ...   
74829  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
74830  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
74831  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
74832  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
74833  http://store.steampowered.com/app/681550/Maze_...          NaT   

                                                    tags  \
0          [Strategy, Action, Indie, Casual, Simulation]   
1          [Strategy, Action, Indie, Casual, Simulation]   
2          [Strategy, Action, Indie, Casual, Simulation]   
3          [Strategy, Action, Indie, Casual, Simulation]   
4          [Strategy, Action, Indie, Casual, Simulation]   
...                                                  ...   
74829                        [Indie, Simulation, Racing]   
74830                        [Indie, Simulation, Racing]   
74831  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
74832  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
74833  [Early Access, Adventure, Indie, Action, Simul...   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/761140/reviews/?...   
2      http://steamcommunity.com/app/761140/reviews/?...   
3      http://steamcommunity.com/app/761140/reviews/?...   
4      http://steamcommunity.com/app/761140/reviews/?...   
...                                                  ...   
74829  http://steamcommunity.com/app/610660/reviews/?...   
74830  http://steamcommunity.com/app/610660/reviews/?...   
74831  http://steamcommunity.com/app/658870/reviews/?...   
74832  http://steamcommunity.com/app/658870/reviews/?...   
74833  http://steamcommunity.com/app/681550/reviews/?...   

                                                   specs  price early_access  \
0                                        [Single-player]   4.99        False   
1                                        [Single-player]   4.99        False   
2                                        [Single-player]   4.99        False   
3                                        [Single-player]   4.99        False   
4                                        [Single-player]   4.99        False   
...                                                  ...    ...          ...   
74829  [Single-player, Steam Achievements, Steam Trad...   1.99        False   
74830  [Single-player, Steam Achievements, Steam Trad...   1.99        False   
74831   [Single-player, Steam Achievements, St

In [24]:
# Se guardan las columnas posibles a utilizar en las consultas en un nuevo dataframe
steamGamesFinal = steam_games[['id','title','genres','price','year']]

steamGamesFinal

id                title      genres  price  year
0      761140  lost summoner kitty      action   4.99  2018
1      761140  lost summoner kitty      casual   4.99  2018
2      761140  lost summoner kitty       indie   4.99  2018
3      761140  lost summoner kitty  simulation   4.99  2018
4      761140  lost summoner kitty    strategy   4.99  2018
...       ...                  ...         ...    ...   ...
74829  610660        russian roads      racing   1.99  2018
74830  610660        russian roads  simulation   1.99  2018
74831  658870  exit 2 - directions      casual   4.99  2017
74832  658870  exit 2 - directions       indie   4.99  2017
74833  681550          maze run vr         NaN   4.99  2017

[74833 rows x 5 columns]

In [ ]:
steamGamesFinal.to_csv('csv_limpios/steam_games.csv', index= False)

In [26]:
steamGamesFinal

id                title      genres  price  year
0      761140  lost summoner kitty      action   4.99  2018
1      761140  lost summoner kitty      casual   4.99  2018
2      761140  lost summoner kitty       indie   4.99  2018
3      761140  lost summoner kitty  simulation   4.99  2018
4      761140  lost summoner kitty    strategy   4.99  2018
...       ...                  ...         ...    ...   ...
74829  610660        russian roads      racing   1.99  2018
74830  610660        russian roads  simulation   1.99  2018
74831  658870  exit 2 - directions      casual   4.99  2017
74832  658870  exit 2 - directions       indie   4.99  2017
74833  681550          maze run vr         NaN   4.99  2017

[74833 rows x 5 columns]

In [27]:
#steamGamesFinal2 = steamGamesFinal.sample(n=25000, random_state=324)

In [28]:
#steamGamesFinal2

id                                              title     genres  \
46032  586570                                  metanet hunter cd     action   
48754  555630               a magical high school girl / 魔法の女子高生  adventure   
51863  497090                              amihailu in dreamland      indie   
68958  239430                            q.u.b.e: director's cut      indie   
55176  383580                                       flesh eaters      indie   
...       ...                                                ...        ...   
45374  587410                  hmm founder's pack - gold edition   strategy   
9139   298641                       brawlhalla - collectors pack      indie   
46614  509696  rocksmith® 2014 edition – remastered – booker ...     casual   
68603  308020  fantasy grounds - call of cthulhu: shadows of ...      indie   
33922  685240                                         paradiddle        NaN   

       price  year  
46032   4.99  2017  
48754   9.99  2016  
51863   0.00  2015  
68958   9.99  2014  
55176   1.99  2016  
...      ...   ...  
45374  79.99  2017  
9139   99.99  2015  
46614   2.99  2017  
68603  19.99  2014  
33922   9.99  2017  

[25000 rows x 5 columns]

In [ ]:
# Se convierte el DataFrame a una tabla de Arrow para posteriormente comprimirlo en parquet
table = pa.Table.from_pandas(steamGamesFinal)

# Se especifica el nombre del archivo Parquet y comprime con snappy
parquet_file = 'datos_parquet/steam_games.parquet'
pq.write_table(table, parquet_file, compression='snappy')

In [ ]:
#sg = pd.read_parquet('datos_parquet/steam_games.parquet')
#sg